cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll

```diff
@@ -1,313 +1,278 @@
-  }
-
-  override final Type getType() {
-    result = var.getType()
-  }
-
-  override final string getUniqueId() {
-    result = var.getUniqueId()
-  }
-}
-
-/**
- * A virtual variable representing all escaped memory accessible by the function,
- * including escaped local variables.
- */
-class UnknownVirtualVariable extends VirtualVariable, TUnknownVirtualVariable {
-  FunctionIR f;
-  
-  UnknownVirtualVariable() {
-    this = TUnknownVirtualVariable(f)
-  }
-  
-  override final string toString() {
-    result = "UnknownVvar(" + f + ")"
-  }
-  
-  override final string getUniqueId() {
-    result = "UnknownVvar(" + f + ")"
-  }
-  
-  override final Type getType() {
-    result instanceof UnknownType
-  }
-  
-  final FunctionIR getEnclosingFunctionIR() {
-    result = f
-  }
-}
-
-private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, IntValue startBitOffset,
+private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, Type type, IntValue startBitOffset,
     IntValue endBitOffset) {
   resultPointsTo(instr.getResultAddressOperand().getDefinitionInstruction(), var, startBitOffset) and
+  type = instr.getResultType() and
   if exists(instr.getResultSize()) then
     endBitOffset = Ints::add(startBitOffset, Ints::mul(instr.getResultSize(), 8))
   else
     endBitOffset = Ints::unknown()
 }
 
-private predicate hasOperandMemoryAccess(MemoryOperand operand, IRVariable var, IntValue startBitOffset,
+private predicate hasOperandMemoryAccess(MemoryOperand operand, IRVariable var, Type type, IntValue startBitOffset,
     IntValue endBitOffset) {
   resultPointsTo(operand.getAddressOperand().getDefinitionInstruction(), var, startBitOffset) and
+  type = operand.getType() and
   if exists(operand.getSize()) then
     endBitOffset = Ints::add(startBitOffset, Ints::mul(operand.getSize(), 8))
   else
     endBitOffset = Ints::unknown()
 }
 
-private newtype TMemoryAccess =
-  TVariableMemoryAccess(IRVariable var, IntValue startBitOffset, IntValue endBitOffset) {
-    hasResultMemoryAccess(_, var, startBitOffset, endBitOffset) or
-    hasOperandMemoryAccess(_, var, startBitOffset, endBitOffset)
+private newtype TMemoryLocation =
+  TVariableMemoryLocation(IRVariable var, Type type, IntValue startBitOffset, IntValue endBitOffset) {
+    hasResultMemoryAccess(_, var, type, startBitOffset, endBitOffset) or
+    hasOperandMemoryAccess(_, var, type, startBitOffset, endBitOffset)
   }
   or
-  TUnknownMemoryAccess(UnknownVirtualVariable uvv) or
-  TTotalUnknownMemoryAccess(UnknownVirtualVariable uvv)
-
-private VariableMemoryAccess getVariableMemoryAccess(IRVariable var, IntValue startBitOffset, IntValue endBitOffset) {
-  result = TVariableMemoryAccess(var, startBitOffset, endBitOffset)
-}
+  TUnknownMemoryLocation(FunctionIR funcIR) or
+  TUnknownVirtualVariable(FunctionIR funcIR)
 
-class MemoryAccess extends TMemoryAccess {
-  string toString() {
-    none()
-  }
+abstract class MemoryLocation extends TMemoryLocation {
```

Comment:
Please add QLDoc for `MemoryLocation`, explaining how it relates to variables.

Reply:
Done